modules/apps/expando/expando-web/src/main/java/com/liferay/expando/web/internal/portlet/ExpandoPortlet.java

```diff
@@ -98,9 +99,48 @@ public void addExpando(
 		long resourcePrimKey = ParamUtil.getLong(
 			actionRequest, "resourcePrimKey");
 
-		String name = ParamUtil.getString(actionRequest, "name");
+		String dataType = ParamUtil.getString(actionRequest, "dataType");
+		String precisionType = ParamUtil.getString(
+			actionRequest, "precisionType");
+
 		int type = ParamUtil.getInteger(actionRequest, "type");
 
+		if (Validator.isNotNull(dataType) &&
+			Validator.isNotNull(precisionType)) {
+
+			if (dataType.equals(ExpandoColumnConstants.TYPE_DECIMAL) &&
+				precisionType.equals(ExpandoColumnConstants.PRECISION_64_BIT)) {
+
+				type = ExpandoColumnConstants.DOUBLE;
+			}
+			else if (dataType.equals(ExpandoColumnConstants.TYPE_DECIMAL) &&
+					 precisionType.equals(
+						 ExpandoColumnConstants.PRECISION_32_BIT)) {
+
+				type = ExpandoColumnConstants.FLOAT;
+			}
+			else if (dataType.equals(ExpandoColumnConstants.TYPE_INTEGER) &&
+					 precisionType.equals(
+						 ExpandoColumnConstants.PRECISION_64_BIT)) {
+
+				type = ExpandoColumnConstants.LONG;
+			}
+			else if (dataType.equals(ExpandoColumnConstants.TYPE_INTEGER) &&
+					 precisionType.equals(
+						 ExpandoColumnConstants.PRECISION_32_BIT)) {
+
+				type = ExpandoColumnConstants.INTEGER;
+			}
+			else if (dataType.equals(ExpandoColumnConstants.TYPE_INTEGER) &&
+					 precisionType.equals(
+						 ExpandoColumnConstants.PRECISION_16_BIT)) {
+
+				type = ExpandoColumnConstants.SHORT;
+			}
```

Comment:
We can extract this long `if-else` to a helper method, which will improve code readability.